Implement L0-fetcher

noreklint opened the pull request:
This PR adds L0-fetcher lambda, whichs runs rclone to sync data from SFTP to AWS S3.

Rclone is added as a lambda layer, which is done by first building rclone locally then pushing the zipped dist as part of the fetcher stack.

Resolves #1

Commit:
Add github templates

Commit:
Add cdk to requirements

Commit:
Add cdk settings

Commit:
Update requirements

Commit:
Implement fetcher

Commit:
Add tests

Commit:
Add Makefile for Rclone lambda layer

Commit:
Build rclone layer im stack

Commit:
Run snapshot tests for stacks

Commit:
Add readme

Commit:
Add ci

Commit:
Update ci to build rclone

Commit:
Use python 3.9

skymandr commented on tox.ini:
```diff
@@ -0,0 +1,41 @@
+[tox]
+envlist = py310,mypy,lint
```
CDK only supports 3.9.

skymandr commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,23 @@
+
+        - name: Install General Dependencies
+          run: |
+            python3.10 -m pip install --upgrade pip
```
See comment on tox.ini

skymandr commented on fetcher/handlers/l0_fetcher.py:
```diff
@@ -0,0 +1,62 @@
+
+
+def get_rclone_config_path(
+    ssm_client: BotoClient,
```
You didn't manage to find the client types either, I take it?

noreklint replied:
Nope not for the `boto.client` function. Other than that boto seems typed

skymandr commented on fetcher/handlers/l0_fetcher.py:
```diff
@@ -0,0 +1,62 @@
+    return cmd
+
+
+def lambda_handler(event: Event, _):
```
are you sure that the second argument doesn't have to be named `context`?

noreklint replied:
I think so, I've tried deploying and testing the lambda and no problems detected

skymandr reviewed (approved):
LGTM, though see comments, in particular regarding python version.

skymandr commented on tox.ini:
```diff
@@ -0,0 +1,41 @@
+commands =
+    mypy --check-untyped-defs --warn-unused-ignores --no-incremental \
+    	--show-error-codes \
+    	fetcher tests
```
borde inte handlers ingå här förresten?

skymandr replied:
ah, nej. ser att du har ett lite annat katalogupplägg än jag (jag har handler i roten). Ditt är nog bra, men borde inte cdk.json uppdateras i så fall, så att den inte packar med stack-koden?

skymandr commented on cdk.json:
```diff
@@ -0,0 +1,40 @@
+      "requirements*.txt",
+      "**/__init__.py",
+      "python/__pycache__",
+      "tests"
```
tox.ini etc?

skymandr commented on tests/fetcher/test_l0_fetcher_stack.py:
```diff
@@ -0,0 +1,18 @@
+
```
```suggestion
```

Commit:
Fix tests

Commit:
Fix github ci

Commit:
Remove stuff from cdk.json

Commit:
Fix lint error

Commit:
Stop using snapshot tests

Snapshot tests using syrupy seems to be really troublesome when combined
with hashed keys/values in AWS templates. It just seems better to do
more fine-grained tests instead...
